File: Sliding Window Median.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers.
     * @return: The median of the element inside the window at each moving
     */
    vector<int> medianSlidingWindow(vector<int> &nums, int k) {
        // write your code here
        vector<int> ret;
        multiset<int, less<int>> min_set;
        multiset<int, greater<int>> max_set;
        for (int i = 0; i < nums.size(); ++i)
        {
            if (i >= k)
            {
                //duplicates
                // min_set.erase(nums[i - k]);
                // max_set.erase(nums[i - k]);
                // Remove the element outside the window.
                //first find from max_set to ensure min_set has the max numbers
                if (max_set.find(nums[i - k]) != max_set.cend()) {
                    max_set.erase(max_set.find(nums[i - k]));
                } else {
                    min_set.erase(min_set.find(nums[i - k]));
                }
            }
            if (max_set.empty() || *max_set.cbegin() < nums[i])
            {
                min_set.emplace(nums[i]);
                if (min_set.size() > max_set.size() + 1)
                {
                    max_set.emplace(*min_set.cbegin());
                    min_set.erase(min_set.begin());
                }
            }
            else
            {
                max_set.emplace(nums[i]);
                if (max_set.size() > min_set.size())
                {
                    min_set.emplace(*max_set.cbegin());
                    max_set.erase(max_set.begin());
                }
            }
            
            if (i >= k - 1)
            {
                ret.emplace_back(min_set.size() > max_set.size() ? *min_set.cbegin() : *max_set.cbegin());
            }
        }
        return ret;
    }
};

File: Sliding Window Median.py
import heapq
class Solution:
    """
    @param nums: A list of integers.
    @return: The maximum number inside the window at each moving.
    """
    def medianSlidingWindow(self, nums, k):
        # write your code here
        minLen, minheap = 0, []
        maxLen, maxheap = 0, []
        mid = 0
        ret = []
        for i in range(len(nums)):

            if not minheap or nums[i] >= minheap[0]:
                heapq.heappush(minheap, nums[i])
                minLen += 1
            else:
                heapq.heappush(maxheap, -1 * nums[i])
                maxLen += 1
            while minLen > maxLen + 1:
                mid = heapq.heappop(minheap)
                minLen -= 1
                heapq.heappush(maxheap, -1 * mid)
                maxLen += 1
            while minLen < maxLen:
                mid = -1 * heapq.heappop(maxheap)
                maxLen -= 1
                heapq.heappush(minheap, mid)
                minLen += 1
            if i >= k - 1:
                if maxLen == minLen:
                    ret.append(-1 * maxheap[0])
                else:
                    ret.append(minheap[0])
            
                if -1 * nums[i - k + 1] in maxheap:
                    maxheap.remove(-1 * nums[i - k + 1])
                    heapq.heapify(maxheap)
                    maxLen -= 1
                elif nums[i - k + 1] in minheap:
                    minheap.remove(nums[i - k + 1])
                    heapq.heapify(minheap)
                    minLen -= 1
        return ret
#note
#without heapq.remove
#not hash to find the index